Email notifications

hannahburkhardt opened the pull request:
send email notification to general practitioners' email addresses when incoming sms is received

hannahburkhardt committed:
send email notification to general practitioners' email addresses when incoming sms is received

hannahburkhardt committed:
provide `false` as default value

mcjustin commented:
@ivan-c  this looks to require some environment config too, but I don't see a PR at https://github.com/uwcirg/isacc-environments/pulls . Can you please aim to include that as you work on the eval system?
https://github.com/uwcirg/isacc-messaging-service/pull/17/files#diff-9facec4a109d7a1b3dc981b379d6e36e74c222e0254254561b8f5c4f08dd8bc6R34-R42

@pbugni  can you please review this?

pbugni commented on isacc_messaging/api/email_notifications.py:
```diff
@@ -0,0 +1,82 @@
+            )
+    except Exception as e:
+        isacc_messaging.audit.audit_entry(
+            f"Email notification could not be sent",
```
i'd suggest you at least include the `recipient_emails` in case the exception text doesn't make it clear, as that's the most likely error source.

pbugni commented on isacc_messaging/api/isacc_record_creator.py:
```diff
@@ -204,6 +225,10 @@ def generate_incoming_message(self, message, time: datetime = None, patient_id=N
         )
+        patient = self.get_patient(patient_id)
+        notify_emails = self.get_general_practitioner_emails(patient)
+        patient_name = " ".join([f"{' '.join(n.given)} {n.family}" for n in patient.name])
```
probably worth adding an accessor or property on Patient for `patient_name` - or we'll end up with formatting code like this being repeated many times.

ivan-c reviewed (approved):
Looks good! May need some minor tweaks to work with CIRG email credentials, but no blockers

mcjustin commented on isacc_messaging/api/isacc_record_creator.py:
```diff
@@ -143,9 +152,21 @@ def get_careplan(self, patient_id):
+            for gp_ref in pt.generalPractitioner:
+                result = HAPI_request('GET', 'Practitioner', gp_ref.id)
+                if result is not None:
+                    gp = Practitioner(result)
```
I've just tested at isacc dev, this line is failing with:
```
Exception: Attempting to instantiate <class 'fhirclient.models.practitioner.Practitioner'> with resource data that defines a resourceType of "Bundle"
```
@pbugni  can you please see if it's an easy fix?

```
messagingservice_1  | [2023-05-09 16:25:52,499] ERROR in app: Exception on /sms [POST]
messagingservice_1  | Traceback (most recent call last):
messagingservice_1  |   File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 2447, in wsgi_app
messagingservice_1  |     response = self.full_dispatch_request()
messagingservice_1  |   File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1952, in full_dispatch_request
messagingservice_1  |     rv = self.handle_user_exception(e)
messagingservice_1  |   File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1821, in handle_user_exception
messagingservice_1  |     reraise(exc_type, exc_value, tb)
messagingservice_1  |   File "/usr/local/lib/python3.7/site-packages/flask/_compat.py", line 39, in reraise
messagingservice_1  |     raise value
messagingservice_1  |   File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1950, in full_dispatch_request
messagingservice_1  |     rv = self.dispatch_request()
messagingservice_1  |   File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1936, in dispatch_request
messagingservice_1  |     return self.view_functions[rule.endpoint](**req.view_args)
messagingservice_1  |   File "/opt/app/isacc_messaging/api/views.py", line 109, in incoming_sms
messagingservice_1  |     result = record_creator.on_twilio_message_received(request.values)
messagingservice_1  |   File "/opt/app/isacc_messaging/api/isacc_record_creator.py", line 319, in on_twilio_message_received
messagingservice_1  |     priority=message_priority
messagingservice_1  |   File "/opt/app/isacc_messaging/api/isacc_record_creator.py", line 229, in generate_incoming_message
messagingservice_1  |     notify_emails = self.get_general_practitioner_emails(patient)
messagingservice_1  |   File "/opt/app/isacc_messaging/api/isacc_record_creator.py", line 161, in get_general_practitioner_emails
messagingservice_1  |     gp = Practitioner(result)
messagingservice_1  |   File "/usr/local/lib/python3.7/site-packages/fhirclient/models/practitioner.py", line 70, in __init__
messagingservice_1  |     super(Practitioner, self).__init__(jsondict=jsondict, strict=strict)
messagingservice_1  |   File "/usr/local/lib/python3.7/site-packages/fhirclient/models/domainresource.py", line 42, in __init__
messagingservice_1  |     super(DomainResource, self).__init__(jsondict=jsondict, strict=strict)
messagingservice_1  |   File "/usr/local/lib/python3.7/site-packages/fhirclient/models/resource.py", line 42, in __init__
messagingservice_1  |     super(Resource, self).__init__(jsondict=jsondict, strict=strict)
messagingservice_1  |   File "/usr/local/lib/python3.7/site-packages/fhirclient/models/fhirabstractresource.py", line 23, in __init__
messagingservice_1  |     .format(self.__class__, jsondict['resourceType']))
messagingservice_1  | Exception: Attempting to instantiate <class 'fhirclient.models.practitioner.Practitioner'> with resource data that defines a resourceType of "Bundle"
```

ivan-c commented:
@hannahburkhardt apologies for taking to long to get back to you

When you get a chance, can you resolve the conflicts on this PR? It also looks to be ~30 commits behind `dev`, but that should also be fixed when conflicts are resolved.

Thank you!!

pbugni committed:
Practitioner query returns bundle.

achen2401 commented on isacc_messaging/api/isacc_record_creator.py:
```diff
@@ -143,9 +152,23 @@ def get_careplan(self, patient_id):
+        emails = []
+        if pt and pt.generalPractitioner:
+            for gp_ref in pt.generalPractitioner:
+                result = HAPI_request('GET', 'Practitioner', gp_ref.id)
```
I _think_  this probably need fixing as it should be extracting the practitioner id value from the `reference` element instead, e.g. see[ example](https://fhir.isacc.dev.cirg.uw.edu/fhir/Patient#L134)
```
                result = HAPI_request('GET', 'Practitioner', gp_ref.split('/')[1])
```

mcjustin replied:
> I _think_ this probably need fixing as it should be extracting the practitioner id value from the `reference` element instead, e.g. see[ example](https://fhir.isacc.dev.cirg.uw.edu/fhir/Patient#L134)
> 
> ```
>                 result = HAPI_request('GET', 'Practitioner', gp_ref.split('/')[1])
> ```

@pbugni  can you please have a look at that?

pbugni replied:
correct fix, thank you @achen2401  .  fwiw, i chose the [-1] slice of the gp_ref split, as we always want the last element in this case (will probably always be 1, but you know how those go...)

pbugni committed:
Handle extra data sent to logging, that can't be JSON encoded (as required by the logging system)

pbugni committed:
Significant debugging cruft included here, till we nail down why emails
are not being sent.  Also includes a number of minor fixes and better
error handling.

pbugni committed:
Silent error in tox only - wrap empty request.get_json() in try/except
to keep passing test.

pbugni committed:
removed excessive debugging output.
adjusted audit entry levels for easier filtering and notification of error / warnings.
removed nop `return None` from the end of several methods.

pbugni committed:
include problematic recipient_emails in error log when a message raises an exception

mcjustin reviewed (approved):
@pbugni  tested on dev looks good, merge when ready
